lib/unparser/preprocessor.rb

```diff
@@ -151,7 +151,44 @@ def chunked_children
 
     end # CollapseStrChildren
 
-    # Preprocessor eliminating unneded dstr nodes
+    # Preprocessor flattening unneeded dstr nesting. This happens when a dstr is generated from
+    # implicit string concatenation, and one of the strings being concatenated is itself a dstr.
+    #
+    # Example: "foo" "bar #{baz}"
+    #
+    # Without this preprocessor, this turns into: "foo#{"bar#{baz}"}"
+    # With this preprocessor, this turns into: "foobar #{baz}"
+    class FlattenImplicitDSTR < self
+
+      # No need to register :dsym, because Ruby doesn't do implicit concatenation of symbols
+      register :dstr
+
+      # Return preprocessor result
+      #
+      # @return [Parser::AST::Node]
+      #
+      # @api private
+      #
+      def result
+        # This should only ever be true for DSTR nodes that are a result of implicit concatenation
+        # (see the comments on this class). Any other DSTR node would have a :begin node as a child.
+        if children.any? && children.all? { |child| [:dstr, :str].include?(child.type) }
```

Comment:
As there is significant of code in the if branch, and only a small expression in the else branch. What about to use the guard form:

```
return node unless children.any? && children.all? { |child| [:dstr, :str].include?(child.type) }
```

This would remove some nesting from the method body making it easier to read and understand.